source/agora/consensus/Quorum.d

```diff
@@ -13,181 +13,476 @@
+unittest
+{
+    auto keys = getKeys(16);
+    auto amounts = Amount(400000000000uL).repeat(16).array;
+    amounts[0] = amounts[$ - 1] = Amount(400000000000uL * 14);
+    auto quorums = buildTestQuorums(amounts, keys);
+    verifyQuorumsSanity(quorums);
+    verifyQuorumsIntersect(quorums);
+    test!"=="(countInclusions(quorums, keys),
+        [16, 2, 7, 7, 6, 5, 5, 6, 6, 8, 5, 9, 7, 3, 4, 16]);
+}
+
+/// 32 nodes where two nodes own 66% of the stake
+unittest
+{
+    auto keys = getKeys(32);
+    auto amounts = Amount(400000000000uL).repeat(32).array;
+    amounts[0] = amounts[$ - 1] = Amount(400000000000uL * 30);
+    auto quorums = buildTestQuorums(amounts, keys);
+    verifyQuorumsSanity(quorums);
+    // verified to work but disabled because it runs slow with a
+    // non-max threshold (~20 seconds)
+    //verifyQuorumsIntersect(quorums);
+    test!"=="(countInclusions(quorums, keys),
+        [31, 9, 3, 6, 5, 6, 6, 3, 4, 6, 4, 4, 2, 6, 4, 5, 6, 6, 6, 4, 8, 9, 5,
+         1, 8, 8, 6, 6, 6, 5, 4, 32]);
+}
+
+/// 64 nodes where two nodes own 66% of the stake
+unittest
+{
+    auto keys = getKeys(64);
+    auto amounts = Amount(400000000000uL).repeat(64).array;
+    amounts[0] = amounts[$ - 1] = Amount(400000000000uL * 62);
+    auto quorums = buildTestQuorums(amounts, keys);
+    verifyQuorumsSanity(quorums);
+    // verified to work but disabled because it runs slow with a
+    // non-max threshold (~20 minutes)
+    //verifyQuorumsIntersect(quorums);
+    test!"=="(countInclusions(quorums, keys),
+        [62, 8, 3, 3, 2, 5, 3, 6, 7, 10, 7, 3, 3, 4, 5, 6, 6, 3, 5, 5, 3, 4, 3,
+         9, 4, 5, 7, 3, 5, 7, 5, 5, 5, 5, 6, 10, 3, 3, 7, 8, 3, 9, 9, 7, 5, 7,
+         5, 8, 6, 3, 4, 2, 2, 1, 7, 9, 5, 5, 6, 6, 4, 5, 5, 62]);
+}
+
+/// 128 nodes where two nodes own 66% of the stake
+unittest
+{
+    auto keys = getKeys(128);
+    auto amounts = Amount(400000000000uL).repeat(128).array;
+    amounts[0] = amounts[$ - 1] = Amount(400000000000uL * 126);
+    auto quorums = buildTestQuorums(amounts, keys);
+    verifyQuorumsSanity(quorums);
+    // not verified to work.
+    //verifyQuorumsIntersect(quorums);
+    test!"=="(countInclusions(quorums, keys),
+        [125, 9, 6, 5, 6, 9, 3, 3, 2, 9, 4, 7, 3, 7, 6, 3, 2, 4, 4, 4, 2, 2, 9,
+        5, 3, 7, 2, 7, 4, 4, 3, 4, 7, 9, 5, 10, 7, 5, 5, 3, 5, 3, 5, 8, 7, 5, 3,
+        2, 2, 6, 6, 3, 6, 3, 4, 4, 8, 8, 3, 5, 4, 12, 8, 5, 2, 11, 4, 5, 6, 4,
+        5, 6, 4, 4, 4, 6, 4, 5, 4, 6, 10, 4, 6, 6, 4, 4, 6, 5, 4, 5, 6, 5, 5, 6,
+        4, 8, 8, 3, 7, 3, 2, 6, 4, 7, 6, 4, 2, 5, 5, 6, 6, 7, 5, 2, 4, 5, 4, 7,
+        3, 8, 9, 4, 6, 4, 6, 4, 4, 123]);
+}
+
+/// Create a shorthash from a 64-byte blob to initialize the rnd generator
+private ulong toShortHash (const ref Hash hash) @trusted
+{
+    import libsodium.crypto_shorthash;
+    import std.bitmanip;
+
+    // generated once with 'crypto_shorthash_keygen'
+    static immutable ubyte[crypto_shorthash_KEYBYTES] PublicKey =
```

Comment:
`PublicKey` is a bad name for this. We should just name it `iv` as a kind of initialization vector.